packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/Helpers.cs

```diff
@@ -11,15 +14,47 @@ internal static class Helpers
         private static readonly string _assemblyLocation = Path.GetDirectoryName(typeof(Helpers).Assembly.Location)!;
 
         public static string GetExpectedFromFile(string? parameters = null)
+        {
+            return File.ReadAllText(GetAssetFilePath(parameters));
+        }
+
+        private static string GetAssetFilePath(string? parameters = null)
         {
             var stackTrace = new StackTrace();
-            var stackFrame = stackTrace.GetFrame(1);
-            var method = stackFrame!.GetMethod();
+            var stackFrame = GetRealMethodInvocation(stackTrace);
+            var method = stackFrame.GetMethod();
             var callingClass = method!.DeclaringType;
             var nsSplit = callingClass!.Namespace!.Split('.');
-            var ns = nsSplit[nsSplit.Length - 1];
+            var ns = nsSplit[^1];
             var paramString = parameters is null ? string.Empty : $"({parameters})";
-            return File.ReadAllText(Path.Combine(_assemblyLocation, ns, "TestData", callingClass.Name, $"{method.Name}{paramString}.cs"));
+            return Path.Combine(_assemblyLocation, ns, "TestData", callingClass.Name, $"{method.Name}{paramString}.cs");
+        }
+
+        private static StackFrame GetRealMethodInvocation(StackTrace stackTrace)
+        {
+            int i = 1;
+            while (i < stackTrace.FrameCount)
+            {
+                var frame = stackTrace.GetFrame(i);
+                if (frame!.GetMethod()!.DeclaringType != typeof(Helpers))
+                {
+                    return frame;
+                }
+                i++;
+            }
+
+            throw new InvalidOperationException($"There is no method invocation outside the {typeof(Helpers)} class in the stack trace");
+        }
+
+        public static Compilation GetCompilationFromFile(string? parameters = null)
```

Comment:
I wonder if this could just be internal to MockHelpers.LoadPlugin. Instead of passing in a Compilation it could be a flag indicating whether there is a customization, or it could be the name of the file containing the customization.

Reply:
we may not let the MockHelpers.LoadPlugin to find the file, because we are finding the file name using the stack trace, getting it inside `MockHelpers.LoadPlugin` method pollutes the stack trace, we will have to introduce some more magic.
Passing the filepath in should be acceptable, but we need to export the `GetAssetFilePath` method.
Also I think doing this makes no difference from we passing in the `Compilation`.
Why is passing in the compilation a bad idea? I think it makes the purpose of this thing very specific, the other solution, this is merely a string.